### PR TITLE
RUN-2954 : Create a new endpoint to expose user/token authorizations

### PR DESCRIPTION
**Pull request by mrdubr**
Provides additional REST API endpoint to query user/token authorizations

**Additional context**
There is some uncertainty around the usage of `kind` and `type` rules in the [ACL policies](https://docs.rundeck.com/docs/manual/document-format-reference/aclpolicy-v10.html).
The assumption made in this API is:
- A `kind` is resource category (e.g. node, job, project)
- A `type` is a resource category (e.g. job, project) with a mandatory `specifier` (e.g. name, id, etc...) that identifies a specific resource in of that type.

**Manual Testing**

***Given the following ACL policies***
```
by:
  group: r1  
context:
  application: 'rundeck'
description: r1 acl
for:
  project:
    - match:
        name: 'P1'
      allow: [read]
  resource:
    - equals:
        kind: system_acl
      allow: [read] # allow modifying system ACL files      
```

```
description: Proj.
context:
  project: 'P1' # all projects
for:
  resource:
    - equals:
        kind: node
      allow: [read] 
  job:
    - allow: [read] 
by:
  group: r1
```

***API Calls***

For an API token that belongs to a user in group/role `r1`:

```
GET http://localhost:4440/api/52/authorizations/application/system_acl?actions=admin&actions=read

{
  "actionAuthorizations": [
    {
      "isAuthorized": false,
      "actionName": "admin"
    },
    {
      "isAuthorized": true,
      "actionName": "read"
    }
  ],
  "resource": {
    "kind": "system_acl"
  },
  "authorizationContext": {
    "type": "application"
  }
}
```

```
GET http://localhost:4440/api/52/authorizations/project/P1/node?actions=admin&actions=read

{
  "actionAuthorizations": [
    {
      "isAuthorized": true,
      "actionName": "read"
    },
    {
      "isAuthorized": false,
      "actionName": "admin"
    }
  ],
  "resource": {
    "kind": "node"
  },
  "authorizationContext": {
    "name": "P1",
    "type": "project"
  }
}
```

```
GET http://localhost:4440/api/52/authorizations/project/P1/node?actions=admin&actions=read

{
  "actionAuthorizations": [
    {
      "isAuthorized": false,
      "actionName": "admin"
    },
    {
      "isAuthorized": true,
      "actionName": "read"
    }
  ],
  "resource": {
    "kind": "node"
  },
  "authorizationContext": {
    "name": "P1",
    "type": "project"
  }
}

```

```
GET http://localhost:4440/api/52/authorizations/project/P1/job/8efdeea8-3c09-40ad-a056-4278ab2a39b1?actions=admin&actions=read

{
  "actionAuthorizations": [
    {
      "isAuthorized": true,
      "actionName": "read"
    },
    {
      "isAuthorized": false,
      "actionName": "admin"
    }
  ],
  "resource": {
    "specifier": "8efdeea8-3c09-40ad-a056-4278ab2a39b1",
    "type": "job"
  },
  "authorizationContext": {
    "name": "P1",
    "type": "project"
  }
}

```

### CODE REVIEW

**Review by smartinellibenedetti — APPROVED** [review]
LGTM, besides that comment, my only suggestion would be to update the CONTEXT_ALLOWED_TYPES with constants (all of them exist in AuthConstants)

File: rundeckapp/grails-app/controllers/rundeck/controllers/AuthorizationsController.groovy [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+            summary = "Get authorizations for a type with specifier",
+            description='''Get authorizations for the supplied set of actions for the subject executing the API call. 
+Evaluation is made in the context of the project for the type with a specifier.
+
+**INCUBATING**: This endpoint is in "incubating" status, and may change.''',
+            tags=['authorization'],
+            parameters = [
+                    @Parameter(name = 'project', description = 'Project Name', required = true, in = ParameterIn.PATH, schema = @Schema(type = 'string')),
+                    @Parameter(name = 'type', description = 'Resource Type', required = true, in = ParameterIn.PATH, schema = @Schema(type = 'string')),
+                    @Parameter(name = 'specifier', description = 'Job Id', required = true, in = ParameterIn.PATH, schema = @Schema(type = 'string')),
```
[the remark this answers is not in the document]
Comment (mrdubr) [reply rundeckapp/grails-app/controllers/rundeck/controllers/AuthorizationsController.groovy:218]: Hmm... I think `specifier` is already pretty generic thus can't think of a way to make it even more generic.

I was considering making it a bit more specific (e.g. `jobId`), but that would prevent reuse of the generic input validation logic, thus decided to keep it as is. Pros and cons.

[the remark this answers is not in the document]
Comment (smartinellibenedetti) [reply rundeckapp/grails-app/controllers/rundeck/controllers/AuthorizationsController.groovy:219]: Sorry, I was talking just about the description itself - like seeing both endpoints with the specifier description of job id gives the impression that both can be used for that, when in fact if the specifier is the job id, it's preferred to use `/authorizations/project/{project}/job/{specifier}`. Saw that in some places the description for specifier is "resource specifier", which could be reused here as well

[the remark this answers is not in the document]
Comment (mrdubr) [reply rundeckapp/grails-app/controllers/rundeck/controllers/AuthorizationsController.groovy:219]: Oh, I see... 
The copy/paste mistake on my side.
Absolutely, in the case of this endpoint -- `/authorizations/project/{project}/{type}/{specifier}`, the `specifier` is not expected to be a jobId! Good catch (as always)!

[the remark this answers is not in the document]
Comment (mrdubr) [reply rundeckapp/grails-app/controllers/rundeck/controllers/AuthorizationsController.groovy]: Fixed: https://github.com/rundeck/rundeck/pull/9507/commits/724763bf2303f8aa26f4841e750b3f07d29299e7

**Pull request merged**